ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -79,25 +94,53 @@ def publish(self):
             self.pub.publish(lane)
 
             # This is needed for visualising in rviz
-            path = Path()
+            path = Marker()
             path.header.frame_id = '/world'
             path.header.stamp = rospy.Time.now()
-            path.poses = map(lambda wp: wp.pose, waypoints)
+            path.ns = "path"
+            path.id = 0
+            path.action = Marker.ADD
+            path.type = Marker.LINE_LIST
+            path.scale.x = 0.1
+            path.color.r = 1.0
+            path.color.a = 1.0
+            path.points = []
+            for wp in waypoints:
+                pos = copy.deepcopy(wp.pose.pose.position)
+                path.points.append(pos)
+                pos = copy.deepcopy(pos)
+                pos.z = wp.twist.twist.linear.x
+                path.points.append(pos)
             self.pub_path.publish(path)
 
-    def pose_cb(self, msg):
-        if self.ego is None or self.ego.header.seq < msg.header.seq:
-            # TODO: Calculate ego's velocity (and maybe acceleration) if needed
-            self.ego = msg
-            self.publish()
+    def get_lookahead(self, start_idx):
+        for i in range(LOOKAHEAD_WPS):
+            idx = (start_idx + i + self.n_waypoints) % self.n_waypoints
+            yield copy.deepcopy(self.waypoints[idx])
+
+    def pose_cb(self, pose):
+        if self.ego is None or self.ego.header.seq < pose.header.seq:
+            self.ego = pose
 
     def waypoints_cb(self, lane):
-        if self.waypoints is None:
+        if self.waypoints is None and lane.waypoints is not None:
             self.waypoints = lane.waypoints
+            self.n_waypoints = len(lane.waypoints)
+            # Unsubscribe so that waypoint loader stops publishing
+            self.base_waypoints_sub.unregister()
```

Comment:
I believe I removed this `unregister` some time ago, since Udacity found a better approach

Reply:
What is Udacity's approach?

Reply:
In this commit we merged their changes:
https://github.com/CarND-EuroBots/CarND-Capstone/commit/911e5f040f8f939ef1cc3246069c18ba011a5553